### ISSUE TITLE
Add optional k_neighbors input parameter to PerplexityBasedNN

### ISSUE THREAD

**Issue by dkobak** [pavlin-policar/openTSNE#156]
Currently `PerplexityBasedNN` uses the standard heuristic 
```
k_neighbors = min(self.n_samples - 1, int(3 * self.perplexity))
```
What do you think about adding an input parameter `k_neighbors='auto'`. The default `'auto'` would use the current behaviour. But e.g. if `perplexity=30, k_neighbors=40`, then that's going to be used as is. Or also `perplexity=30, k_neighbors=200`.

This came up in a project I am doing with a rotation MSc student, so I was wondering if it would make sense to allow this flexibility to PerplexityBasedNN. I think it'd be nice because currently `3` in `3 * self.perplexity` is a hard-coded constant which is not great.

**Comment by pavlin-policar** [comment]
I have thought about this in the context of scanpy, where this behaviour might be desirable, but it's not really clear to me how this would play with calculating the Gaussian kernel widths. Intuitively, I don't think it should really affect them much, because the perplexity of the distribution can be calculated on any number of samples. I would assume that just the accuracy would suffer if we used too few samples i.e. the kernel bandwidths might be estimated incorrectly. Not that I think this would be problematic, since the uniform kernel works just as well.

I think this is something we will have to add to accomodate the scanpy use-case, but we'll have to figure out what parameter value combinations are reasonable. 3\*k is hard-coded because it makes sense. Neighbors further than the 3*k nearest neighbors will be assigned a practically zero value, so parameter combinations like `perplexity=30, k=500` should be heavily discouraged. On the other hand, running something like `perplexity=500, k=30` would make no sense and would lead to incorrect results.

Off the top of my head, something like
```python
if k > 3 * perplexity:
    warn
elif k < perplexity * 2:
    error or at least warn
```

**Comment by dkobak** [comment]
It's mathematically impossible to obtain perplexity larger than the number of neighbors. Uniform kernel yields maximum perplexity, and that's equal to the number of neighbors. So if `k > perplexity`, there should be an error. We can give a warning if `k > 3 * perplexity`, why not.

**Comment by pavlin-policar** [comment]
Hmm, you're right, but I don't really understand why. I've played around with this a little, and, as you said, the moment `k < perplexity`, the bandwidths become basically zero. What's the intuition behind this?

But, basically, yes, that would definitely warrant an error!

**Comment by dkobak** [comment]
I think the Gaussian variance sigma should go to infinity in this limit, not to zero.

It's because the uniform distribution has the max entropy among discrete distributions with the same support. Hence it has the max perplexity. And the perplexity of the uniform distribution defined on k objects is k.

**Comment by pavlin-policar** [comment]
Closed via #157.
